File: 2016/day05/solution.swift
#!/usr/bin/env swift

import Foundation

func shell(_ args: String...) -> String? {
    let task = Process()
    task.launchPath = "/usr/bin/env"
    task.arguments = args

    let pipe = Pipe()
    task.standardOutput = pipe
    task.launch()

    let data = pipe.fileHandleForReading.readDataToEndOfFile()
    let output = String(data: data, encoding: .utf8)

    task.waitUntilExit()
    return output
}

func md5(_ input: String) -> String? {
    guard let md5Response = shell("md5", "-s", input) else { return nil }
    return md5Response.components(separatedBy: " ").last
}

/*
func md5(_ string: String) -> String {
    var digest = Array<UInt8>(repeating:0, count:Int(CC_MD5_DIGEST_LENGTH))
    CC_MD5(string, CC_LONG(string.lengthOfBytes(using: .utf8)), &digest)
    let hexString = digest.map { String(format: "%02hhx", $0) }.joined()

    return hexString
}
*/

// ------------------------------------------------------------------
// MARK: - "MAIN()"

print("Unable to shell out enough times, quickly enough, to solve this.")
print("Unable to determine how to import CommonCrypto into a script to perform MD5 in code.")

/*
let input = "abc"
// let input = "wtnhxymk"

var passwordChars = [Character]()
var hashSuffix = 0
while passwordChars.count < 2 {
    if let hash = md5("\(input)\(hashSuffix)") {
        if hash.hasPrefix("00000") {
            print("\(hashSuffix) -> \(hash)")
            let passwordCharIndex = hash.index(hash.startIndex, offsetBy: 5)
            passwordChars.append(hash[passwordCharIndex])
        }
    }
    hashSuffix = hashSuffix + 1

    if hashSuffix % 1000 == 0 {
        print("Mod 1k: \(hashSuffix)")
    }
}

print(md5("abc3231929")!)
print(md5("abc5017308")!)
*/

File: 2016/day05/solution.rb
#!/usr/bin/env ruby

require 'digest/md5'

input = 'wtnhxymk'
suffix_index = 0

password_chars = []
password = '_'*8

while password =~ /_/ do
    md5 = Digest::MD5.hexdigest(input + suffix_index.to_s)
    if md5 =~ /^00000/ then
        p "#{input}#{suffix_index} -> #{md5} [#{md5[5]}, #{md5[6]}]"

        password_chars << md5[5] if password_chars.length < 8

        index = md5[5].to_i unless ('a'..'z').include?(md5[5])
        if (0..7).include?(index) then
          password[index] = md5[6] if password[index] == '_'
          p password
        end
    end
    suffix_index += 1
end

p '* Password: ' + password_chars.join('')
p '** Password: ' + password
